book/emails.rst

```diff
@@ -0,0 +1,320 @@
+         */
+        private $emails;
+
+        /**
+         * @ORM\Column(type="string", length=255)
+         */
+        private $firstName;
+
+        /**
+         * @ORM\Column(type="string", length=255)
+         */
+        private $lastName;
+
+        public function getClass()
+        {
+            return 'Acme\DemoBundle\Entity\Applicant';
+        }
+
+        public function getEmailFields()
+        {
+            return array('email');
+        }
+
+        public function getId()
+        {
+            return $this->id;
+        }
+
+        public function getEmails()
+        {
+            return $this->emails;
+        }
+
+        public function getFirstName()
+        {
+            return $this->firstName;
+        }
+
+        public function getLastName()
+        {
+            return $this->lastName;
+        }
+    }
+
+.. _book-emails-owner-provider:
+
+Implementing the ``EmailOwnerProviderInterface``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to make the application able to find the owner of a certain e-mail address, you have to
+create a provider that implements the
+:class:`Oro\\Bundle\\EmailBundle\\Entity\\Provider\\EmailOwnerProviderInterface`. This interface
+contains two methods:
+
+``getEmailOwnerClass()``
+    This is the class of the e-mail owner entity (the class implementing the
+    ``EmailOwnerInterface`` which is the ``Applicant`` class in the example above).
+
+``findEmailOwner()``
+    Returns an entity that is the owner of an e-mail address or ``null`` if no such owner exists.
+    The returned object must be an instance of the class specified by ``getEmailOwnerClass()``.
+
+The provider class should then look like this:
+
+.. code-block:: php
+    :linenos:
+
+    // src/Acme/DemoBundle/Entity/Provider/EmailOwnerProvider.php
+    namespace Acme\DemoBundle\Entity\Provider;
+
+    use Acme\DemoBundle\Entity\Applicant;
+    use Acme\DemoBundle\Entity\ApplicantEmail;
+    use Doctrine\ORM\EntityManager;
+    use Oro\Bundle\EmailBundle\Entity\Provider\EmailOwnerProviderInterface;
+
+    class EmailOwnerProvider implements EmailOwnerProviderInterface
+    {
+        public function getEmailOwnerClass()
+        {
+            return 'Oro\Bundle\UserBundle\Entity\User';
+        }
+
+        public function findEmailOwner(EntityManager $em, $email)
+        {
+
+            $applicantRepository = $em->getRepository('AcmeDemoBundle:Applicant');
+            $applicant = $applicantRepository->findOneBy(array('email' => $email));
+
+            if (null === $applicant) {
+                $emailRepository = $em->getRepository('AcmeDemoBundle:Email');
+                $emailEntity = $emailRepository->findOneBy(array('email' => $email));
+
+                if (null !== $emailEntity) {
+                    $applicant = $emailEntity->getEmailOwner();
+                }
+            }
+
+            return $applicant;
```

Comment:
The implementation of findEmailOwner should be changed like so:
        $applicantEmailRepo = $em->getRepository('AcmeDemoBundle:ApplicantEmail');
        $applicantEmail = $applicantEmailRepo->findOneBy(array('email' => $email));

        if (null !== $applicantEmail) {
            return $applicantEmail->getEmailOwner();
        }

        return null;